pkg/monitoragent/monitoragent_linux.go

```diff
@@ -0,0 +1,217 @@
+	errUnexpectedEvent      = errors.New("unexpected event type for MessageTypeAgent")
+)
+
+// isCtxDone is a utility function that returns true when the context's Done()
+// channel is closed. It is intended to simplify goroutines that need to check
+// this multiple times in their loop.
+func isCtxDone(ctx context.Context) bool {
+	select {
+	case <-ctx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+type monitorAgent struct {
+	lock.Mutex
+	models.MonitorStatus
+
+	ctx              context.Context
+	perfReaderCancel context.CancelFunc
+
+	// listeners are external cilium monitor clients which receive raw
+	// gob-encoded payloads
+	listeners map[listener.MonitorListener]struct{}
+	// consumers are internal clients which receive decoded messages
+	consumers map[consumer.MonitorConsumer]struct{}
+}
+
+func (a *monitorAgent) AttachToEventsMap(int) error {
+	return nil
+}
+
+func (a *monitorAgent) SendEvent(typ int, event interface{}) error {
+	if a == nil {
+		return errMonitorAgentNotSetup
+	}
+
+	// Two types of clients are currently supported: consumers and listeners.
+	// The former ones expect decoded messages, so the notification does not
+	// require any additional marshalling operation before sending an event.
+	// Instead, the latter expect gob-encoded payloads, and the whole marshalling
+	// process may be quite expensive.
+	// While we want to avoid marshalling events if there are no active
+	// listeners, there's no need to check for active consumers ahead of time.
+
+	a.notifyAgentEvent(typ, event)
+
+	// do not marshal notifications if there are no active listeners
+	if !a.hasListeners() {
+		return nil
+	}
+
+	// marshal notifications into JSON format for legacy listeners
+	if typ == api.MessageTypeAgent {
+		msg, ok := event.(api.AgentNotifyMessage)
+		if !ok {
+			return errUnexpectedEvent
+		}
+		var err error
+		event, err = msg.ToJSON()
+		if err != nil {
+			return fmt.Errorf("unable to JSON encode agent notification: %w", err)
+		}
+	}
+
+	var buf bytes.Buffer
+	if err := buf.WriteByte(byte(typ)); err != nil {
+		return fmt.Errorf("unable to initialize buffer: %w", err)
+	}
+	if err := gob.NewEncoder(&buf).Encode(event); err != nil {
+		return fmt.Errorf("unable to gob encode: %w", err)
+	}
+
+	p := payload.Payload{Data: buf.Bytes(), CPU: 0, Lost: 0, Type: payload.EventSample}
+	a.sendToListeners(&p)
+
+	return nil
+}
+
+func (a *monitorAgent) RegisterNewListener(newListener listener.MonitorListener) {
+	if a == nil || newListener == nil {
+		return
+	}
+
+	a.Lock()
+	defer a.Unlock()
+
+	if isCtxDone(a.ctx) {
+		log.Debug("RegisterNewListener called on stopped monitor")
+		newListener.Close()
+		return
+	}
+
+	version := newListener.Version()
+	switch newListener.Version() { //nolint:exhaustive // the only other case is unsupported which is covered by default
+	case listener.Version1_2:
+		a.listeners[newListener] = struct{}{}
+	default:
+		newListener.Close()
```

Comment:
same here, exposing close behavior in an exposed register function could result in unexpected behavior by function name

Reply:
Same as https://github.com/microsoft/retina/pull/432#discussion_r1649442249